src/samplers/cpu/mod.rs

```diff
@@ -149,6 +121,73 @@ impl Sampler for Cpu {
 }
 
 impl Cpu {
+    #[cfg(feature = "bpf")]
+    fn perf_enabled(&self) -> bool {
+        if self.sampler_config().perf_events() {
+            for statistic in self.sampler_config().statistics() {
+                if statistic.perf_config().is_some() {
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    fn initialize_perf(&mut self) -> Result<(), failure::Error> {
+        #[cfg(feature = "bpf")]
+        {
+            if self.enabled() && self.perf_enabled() {
+                debug!("initializing perf");
+                let code = include_str!("perf.c");
+                let mut perf = bcc::BPF::new(code)?;
+
+                for statistic in self.sampler_config().statistics() {
+                    if let Some((name, event)) = statistic.perf_config() {
+                        bcc::PerfEvent::new()
+                            .handler(&format!("f_{}", name))
+                            .event(event)
+                            .sample_period(Some(SAMPLE_PERIOD))
+                            .attach(&mut perf)?;
+                    }
+                }
+
+                self.perf = Some(Arc::new(Mutex::new(BPF { inner: perf })))
+            }
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "bpf")]
+    fn sample_perf(&self) -> Result<(), std::io::Error> {
+        if self.perf_last.lock().unwrap().elapsed() >= self.general_config().window() {
```

Comment:
Ah - so this shouldn't be done this way. For the other BPF samplers, we are aggregating in a histogram, but for the perf event ones, we're using raw counters. The idea here is that for BPF, we read the BPF histogram into the Rezolus histogram once per window (60s by default). For perf counters, we want to actually read those counters every time we sample for the given sampler.